writeaheadlog_integration_test.go

```diff
@@ -1,556 +1,487 @@
-	var updates []Update
-	for i := 0; i < len(ca.countdown); i++ {
-		ca.countdown[i]++
-		updates = append(updates, Update{
-			Name:    addCountName,
-			Version: "1.0.0",
-			Instructions: addCountUpdate{
-				index:    uint64(i),
-				newCount: ca.countdown[i],
-			}.marshal(),
-		})
-	}
+		// ncs is the new checkum of the silo and the corresponding file
+		ncs checksum
 
-	// Corner case - if there are no updates, because this is the first time we
-	// add to the count, just write out the file piece.
-	if len(updates) == 0 {
-		writeBytes := make([]byte, 8)
-		_, err := ca.file.WriteAt(writeBytes, 8*int64(len(ca.countdown)))
-		if err != nil {
-			return err
-		}
-		err = ca.file.Sync()
-		if err != nil {
-			return err
-		}
-		ca.countdown = append(ca.countdown, 0)
-		return nil
+		// ncso is the offset of the new checksum
+		ncso int64
 	}
+)
 
-	// Create the WAL transaction.
-	tx, err := ca.wal.NewTransaction(updates)
-	if err != nil {
-		return err
+// newSilo creates a new silo of a certain length at s specific offset in the
+// file
+func newSilo(offset int64, length int, deps dependencies, f file) *silo {
+	if length == 0 {
+		panic("numbers shouldn't be empty")
 	}
-	// Perform the setup write on the file.
-	writeBytes := make([]byte, 8)
-	_, err = ca.file.WriteAt(writeBytes, 8*int64(len(ca.countdown)))
-	if err != nil {
-		return err
-	}
-	err = ca.file.Sync()
-	if err != nil {
-		return err
-	}
-	// Signal completed setup and then wait for the commitment to finish.
-	errChan := tx.SignalSetupComplete()
-	err = <-errChan
-	if err != nil {
-		return err
-	}
-	// Apply the updates.
-	ca.countdown = append(ca.countdown, 0)
-	err = ca.applyUpdates(updates)
-	if err != nil {
-		return err
+
+	return &silo{
+		offset:  offset,
+		numbers: make([]uint32, length, length),
+		f:       f,
+		deps:    deps,
 	}
-	err = tx.SignalUpdatesApplied()
-	if err != nil {
-		return err
+}
+
+// marshal marshals a siloUpdate
+func (su siloUpdate) marshal() []byte {
+	data := make([]byte, 28+2*checksumSize)
+	binary.LittleEndian.PutUint64(data[0:8], uint64(su.offset))
+	binary.LittleEndian.PutUint32(data[8:12], su.number)
+	binary.LittleEndian.PutUint64(data[12:20], uint64(su.silo))
+	binary.LittleEndian.PutUint64(data[20:28], uint64(su.ncso))
+	copy(data[28:28+checksumSize], su.ocs[:])
+	copy(data[28+checksumSize:], su.ncs[:])
+	return data
+}
+
+// unmarshal unmarshals a siloUpdate from data
+func (su *siloUpdate) unmarshal(data []byte) {
+	if len(data) != 28+2*checksumSize {
+		panic("data has wrong size")
+	}
+	su.offset = int64(binary.LittleEndian.Uint64(data[0:8]))
+	su.number = binary.LittleEndian.Uint32(data[8:12])
+	su.silo = int64(binary.LittleEndian.Uint64(data[12:20]))
+	su.ncso = int64(binary.LittleEndian.Uint64(data[20:28]))
+	copy(su.ocs[:], data[28:28+checksumSize])
+	copy(su.ncs[:], data[28+checksumSize:])
+	return
+}
+
+// newUpdate create a WAL update from a siloUpdate
+func newUpdate(su *siloUpdate) Update {
```

Comment:
might want to make this a method of the `siloUpdate` object

Reply:
will do